overlord/snapstate/refreshhints.go

```diff
@@ -121,3 +135,111 @@ func (r *refreshHints) Ensure() error {
 	}
 	return r.refresh()
 }
+
+func refreshHintsFromCandidates(st *state.State, updates []*snap.Info, ignoreValidationByInstanceName map[string]bool, deviceCtx DeviceContext) ([]*RefreshHint, error) {
+	if ValidateRefreshes != nil && len(updates) != 0 {
+		userID := 0
+		var err error
+		updates, err = ValidateRefreshes(st, updates, ignoreValidationByInstanceName, userID, deviceCtx)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	hints := make([]*RefreshHint, len(updates))
+	for _, update := range updates {
+		var snapst SnapState
+		if err := Get(st, update.InstanceName(), &snapst); err != nil {
+			return nil, err
+		}
+		// XXX: is this ok?
+		flags := snapst.Flags
+		flags.IsAutoRefresh = true
+
+		flags, err := ensureInstallPreconditions(st, update, flags, &snapst)
+		if err != nil {
+			// XXX: do we need to handle refreshAll here?
```

Comment:
not sure what this XXX refers to?

Reply:
It's referring to the old code where this originally came from and would handle refreshAll; it would ignore single failures and try to update. Again mostly a note to myself to double check this.